core, xeth: chain reorg move missing transactions to transaction pool

obscuren opened the pull request:
This PR addresses functionality to the `ChainManager` so that missing transaction that may occur during chain forks get inserted back in to the `TransactionPool`. It calculated the difference between old and new chain and picks out transactions that never ended up in the new chain.

**NOTE: For reference, this PR seen quite some iterations. The following text is OLD**

***

Added an new transaction type used by the transaction pool `poolTx`.
This struct hold information about the transaction state:

* amount of post tries
* added timestamp
* last post timestamp

Whenever a transaction is being submitted by the user the transaction is
flagged as owned. When an owned transaction sits in the pool for too
long (39 seconds) it will be posted again.

A future PR could check the amount of post tries and will post back a
failure event. An failure event could potentially be posted back to a
frontend connected to the `geth` instance and deletes from the pool. A
frontend could ask the user for a higher gas price or some other action.

***

This PR also adds back any potential missing transactions during chain reorganisations and adds functionality to `types.Transactions` to split of `Difference`s between two slices of `Transaction`s. 

Closes #1668

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Sun Aug 16 13:22:10 UTC 2015_

Gustav-Simonsson commented on core/chain_manager.go:
```diff
@@ -663,6 +663,82 @@ func (self *ChainManager) InsertChain(chain types.Blocks) (int, error) {
+	)
+
+	// first reduce whoever is higher bound
+	if oldBlock.NumberU64() > newBlock.NumberU64() {
```
Technically the longer chain is the one with higher total difficulty. While unlikely to happen, it's possible we could see a reorg to a chain with higher TD but lower block number.

Gustav-Simonsson commented on core/chain_manager.go:
```diff
@@ -663,6 +663,82 @@ func (self *ChainManager) InsertChain(chain types.Blocks) (int, error) {
+			commonBlock = oldBlock
+			break
+		}
+		newChain = append(newChain, newBlock)
```
Will this not append newBlock a second time if it was already appended at line 691?

Gustav-Simonsson commented on core/chain_manager.go:
```diff
@@ -663,6 +663,82 @@ func (self *ChainManager) InsertChain(chain types.Blocks) (int, error) {
+		}
+		newChain = append(newChain, newBlock)
+
+		oldBlock, newBlock = self.GetBlock(oldBlock.ParentHash()), self.GetBlock(newBlock.ParentHash())
```
If old chain was reduced above, it'll set oldBlock to it's parent in the for loop post statement. Then here it's set to it's grandparent. Looks like it'll skip adding the parent's txs to deletedTxs in this case.

Gustav-Simonsson commented on core/events.go:
```diff
@@ -36,6 +36,9 @@ type NewBlockEvent struct{ Block *types.Block }
 // NewMinedBlockEvent is posted when a block has been imported.
 type NewMinedBlockEvent struct{ Block *types.Block }
 
+// RemovedTransactionEvent is posted when a reord happens
```
reord -> reorg

Gustav-Simonsson commented on core/transaction_pool.go:
```diff
@@ -45,10 +46,21 @@ var (
 const (
-	maxQueued = 64 // max limit of queued txs per address
+	maxQueued  = 64               // max limit of queued txs per address
+	resendTime = 39 * time.Second // every ~3 blocks
```
If 3 blocks is desired this should be 45s since resendTime is compared against real time and it looks like we get 15s real time from the 13s target time.

Gustav-Simonsson commented on core/transaction_pool.go:
```diff
@@ -100,6 +112,10 @@ func (pool *TxPool) eventLoop() {
 		case GasPriceChanged:
 			pool.minGasPrice = ev.Price
+		case RemovedTransactionEvent:
+			pool.mu.Unlock()
```
Hmm, should unlock/lock be switched here?

obscuren replied:
No, within `AddTransaction` it's being locked, that's why it unlocks there

Gustav-Simonsson commented on core/types/transaction.go:
```diff
@@ -276,14 +276,32 @@ func (tx *Transaction) String() string {
 	return enc
 }
 
+// Difference sets s to the difference of a to b. Expects s to be empty
```
Should it really be a method if s must be empty?

obscuren replied:
Yes, I think you're right. I was pondering the same thing

karalabe commented:
I don't know if this PR addresses an issue related to missing txns due to chain reorgs, but if not please see (https://github.com/ethereum/go-ethereum/issues/1746). Essentially if a chain reorg obsoletes a previously successful transaction, we should also either delete it from the database or make sure it's not happily returned as mined if it "went missing".

obscuren commented:
Adding deletion of missing (due to chain reorg) transactions and receipts

obscuren commented:
@karalabe @Gustav-Simonsson PTAL

Gustav-Simonsson commented on core/chain_manager.go:
```diff
@@ -731,20 +731,26 @@ func (self *ChainManager) InsertChain(chain types.Blocks) (int, error) {
-// diff takes two blocks, an old chain and a new chain and will reconstruct the blocks and inserts them
-// to be part of the new canonical chain.
-func (self *ChainManager) diff(oldBlock, newBlock *types.Block) (types.Blocks, error) {
+// reorgs takes two blocks, an old chain and a new chain and will reconstruct the blocks and inserts them
```
English nitpicking: this mixes [simple future tense](http://www.edufind.com/english-grammar/simple-future-tense/) and [simple present tense](http://www.edufind.com/english-grammar/simple-present-tense/)

Would recommend to always go with present tense for source code comments describing code after the comment and past tense for comments describing code prior to it.

fjl replied:
http://2.bp.blogspot.com/_REkgXByDyuU/S7OQdXHzpvI/AAAAAAAAGyY/AwF-R2Zy5C8/s1600/grammar-nazi.jpg

Gustav-Simonsson commented on core/types/transaction.go:
```diff
@@ -272,14 +272,34 @@ func (tx *Transaction) String() string {
 type Transactions []*Transaction
 
-func (s Transactions) Len() int      { return len(s) }
+// Len returns the length of s
```
comment can be removed, code is self-explanatory here.

obscuren replied:
All exported functions should be documented regardless of their simplicity

Gustav-Simonsson commented on core/chain_manager.go:
```diff
@@ -800,8 +797,18 @@ func (self *ChainManager) merge(oldBlock, newBlock *types.Block) error {
+	}
+
+	var diff types.Transactions
+	types.TxDifference(deletedTxs, addedTxs)
```
```types.TxDifference``` returns a new slice as the set different of a to b, but does not modify a or b. Looks like it's return should be assigned here.

obscuren replied:
Thanks. This changes from **modify target** to single function. I didn't update to the new format. Good catch @Gustav-Simonsson

Gustav-Simonsson commented:
:+1:

karalabe commented on core/types/transaction.go:
```diff
@@ -272,14 +272,34 @@ func (tx *Transaction) String() string {
+		if _, ok := m[tx.Hash()]; !ok {
+			t = append(t, tx)
+		}
+	}
```
This segment could be simplified a lot by using a `map[common.Hash]bool`. Then the first iteration would add `m[tx.Hash()] = true`, and the second could simply check `if !m[tx.Hash()]`.

Also I think I'd prefer `t` to be initialized to `t := make(types.Transactions, 0, len(a))`. This way it would be only a single memory alloc and not multiple ones (exponential growth needs some time to kick in).

karalabe replied:
Also maybe rename `m` to `remove` and `t` to `keep`. It's not really nice to use one letter variables, especially since they seem quire arbitrary to me.

obscuren replied:
I'll address the initialisation.

Don't see your point on making anything simpler by changing the ptr to a bool.

karalabe replied:
Then use struct{}, I don't think its good coding style to put some data into a map just to represent presence.

obscuren replied:
Please elaborate on the "good coding" style. I agree with to `struct{}`

bas-vk commented on core/chain_manager.go:
```diff
@@ -801,8 +798,18 @@ func (self *ChainManager) merge(oldBlock, newBlock *types.Block) error {
+	// receipts that were created in the fork must also be deleted
+	for _, tx := range deletedTxs {
+		DeleteReceipt(self.chainDb, tx.Hash())
+		DeleteTransaction(self.chainDb, tx.Hash())
```
Will this remove transactions which are part of the old chain/fork and part of the new chain from the database?

karalabe replied:
Nice catch, that should have been `range diff` imho.

obscuren replied:
It removes transactions that were present in `old chain` but not in `new chain` but leaves `new chain` txs untouched.

karalabe replied:
Why would it leave new ones untouched? If the old and hew hash are the same, won't it delete the freshly inserted stuff too?

obscuren replied:
Difference between `deleted` **to** `added`, not difference **between**. Perhaps `TxDifference` should be renamed.

karalabe replied:
The `diff` is correct, the issue is in the next line. You iterate `deletedTxs`, not the freshly calculated `diff`. So you will delete everything that was reorganized, irrespective whether it was added back or not. Unless I really misunderstand something.

obscuren replied:
OI! :+1:

bas-vk replied:
Ok, that clarifies it.

karalabe commented on core/chain_manager_test.go:
```diff
@@ -483,22 +485,71 @@ func TestInsertNonceError(t *testing.T) {
-	genesis := GenesisBlock(0, db)
-	_, err := NewChainManager(genesis, db, db, db, thePow(), &mux)
-	if err != nil {
+func TestChainReorgMissingTransactions(t *testing.T) {
```
This test is wrong. The two transactions created in the two forks are not the same. Signing introduces a randomness into the signature, causing the two transaction to have different signatures, hence different hashes. If you were to define one single transaction

```
	txn, _ := types.NewTransaction(0, addr2, big.NewInt(1000), params.TxGas, nil, nil).SignECDSA(key1)
```

and inject that into both chains, your tests will fail, per the issue found by @bas-vk .

karalabe replied:
Hmm, sorry, it was my half ass implementation that failed, however, the test is still wrong in that they are not the same transactions. Maybe it would help to have separate tests to test removal, and override?

karalabe replied:
A more complete test imho would be to have three transactions in total. A, B included in the first chain, B, C included in the second chain. Then you could test all three reorg scenarios -> A drops, B stays, C appears

obscuren replied:
The point is not so much to have the same transaction inserted in to the chain, rather it shouldn't. The test is to make sure the transaction is removed from the database.

obscuren replied:
In fact the transactions **SHOULD NOT** be the same.

obscuren replied:
The confusion probably stems from the fact that I'm using the same key and the reliance of the randomness introduces due to the signing. This is because the ` WriteGenesisBlockForTesting` method only accepts 1 state object.

bas-vk commented:
:+1:

karalabe commented:
To fix the failing tests, drop this diff into a file and then do `git apply <diff>`:

```
diff --git a/cmd/geth/js_test.go b/cmd/geth/js_test.go
index 2fd5a53..1f5b28e 100644
--- a/cmd/geth/js_test.go
+++ b/cmd/geth/js_test.go
@@ -92,7 +92,7 @@ func testREPL(t *testing.T, config func(*eth.Config)) (string, *testjethre, *eth
 
        db, _ := ethdb.NewMemDatabase()
 
-       core.WriteGenesisBlockForTesting(db, common.HexToAddress(testAddress), common.String2Big(testBalance))
+       core.WriteGenesisBlockForTesting(db, core.GenesisAccount{common.HexToAddress(testAddress), common.String2Big(testBalance)})
        ks := crypto.NewKeyStorePlain(filepath.Join(tmp, "keystore"))
        am := accounts.NewManager(ks)
        conf := &eth.Config{
diff --git a/common/natspec/natspec_e2e_test.go b/common/natspec/natspec_e2e_test.go
index ea28b45..57f81f6 100644
--- a/common/natspec/natspec_e2e_test.go
+++ b/common/natspec/natspec_e2e_test.go
@@ -134,7 +134,7 @@ func testEth(t *testing.T) (ethereum *eth.Ethereum, err error) {
 
        db, _ := ethdb.NewMemDatabase()
        // set up mock genesis with balance on the testAddress
-       core.WriteGenesisBlockForTesting(db, common.HexToAddress(testAddress), common.String2Big(testBalance))
+       core.WriteGenesisBlockForTesting(db, core.GenesisAccount{common.HexToAddress(testAddress), common.String2Big(testBalance)})
 
        // only use minimalistic stack with no networking
        ethereum, err = eth.New(&eth.Config{
diff --git a/eth/helper_test.go b/eth/helper_test.go
index 3a799e6..034751f 100644
--- a/eth/helper_test.go
+++ b/eth/helper_test.go
@@ -33,7 +33,7 @@ func newTestProtocolManager(blocks int, generator func(int, *core.BlockGen), new
                evmux       = new(event.TypeMux)
                pow         = new(core.FakePow)
                db, _       = ethdb.NewMemDatabase()
-               genesis     = core.WriteGenesisBlockForTesting(db, testBankAddress, testBankFunds)
+               genesis     = core.WriteGenesisBlockForTesting(db, core.GenesisAccount{testBankAddress, testBankFunds})
                chainman, _ = core.NewChainManager(db, pow, evmux)
                blockproc   = core.NewBlockProcessor(db, pow, chainman, evmux)
        )
```

Though I've merged your PR into a freshest develop, so it might crap itself with some merge conflict, but then you can also do a manual patch.

karalabe commented:
I made a patch commit against your PR. Wanted to open a PR for your PR, but I have a different develop as a base, so it got full of just as this PR's not rebased, etc. Long story short, here are a few of my suggested test fixes and modifications. You can cherry pick them into this PR if you wish so. It however does depend on a tiny PR I opened to extend @fjl's chain maker with difficulty modifications (https://github.com/ethereum/go-ethereum/pull/1815).

https://github.com/karalabe/go-ethereum/commit/804ac1cadbde81e818f5772875db7dff98b8ba18

Beside containing the build fixes in the tests that I already mentioned in my previous comment, I've just extended the reorg test to contain a few more corner cases. They are all handled correctly by the implementation (at least the tests pass) but they cover a few more scenarios:
 - Deleting transactions from both old blocks, as well as from the block being reorged.
 - Moving shared transactions from both old blocks, as well as ffrom the block being reorged.
 - Adding new transactions in 1) old blocks, 2) block being reorged, 3) future blocks.

The change (requiring the chain maker PR) also forces block number 3 to be forked by the new block 3 due to high TD. Previously the reorg was only detected at block 4 (old head + 1), making it impossible to test any operations happening within the reorged block.

karalabe commented on core/transaction_pool.go:
```diff
@@ -245,16 +278,16 @@ func (self *TxPool) add(tx *types.Transaction) error {
+		self.queue[from] = make(map[common.Hash]*poolTx)
 	}
-	self.queue[from][hash] = tx
+	self.queue[from][hash] = &poolTx{tx, owned, time.Now(), time.Time{}, 0}
```
Could you write out the field names too here? (i.e. of `poolTx`). With 2 fields it's ok-ish, but 5 is pushing it and can lead to odd issues if we reorganize the order of the fields, add new ones, etc.

karalabe commented on core/transaction_pool.go:
```diff
@@ -100,6 +112,10 @@ func (pool *TxPool) eventLoop() {
 			pool.minGasPrice = ev.Price
+		case RemovedTransactionEvent:
+			pool.mu.Unlock()
+			pool.AddTransactions(ev.Txs)
```
There's a small logical omission here that might have implications.  `AddTransactions` doesn't keep track of transaction ownership (https://github.com/ethereum/go-ethereum/pull/1669/files#diff-d37f7abd23589257d36b8fa71ac686faR331). I.e. even if the transaction was originally owned by this node, that meta info is lost.

However, given that the transaction was already integrated once, we might assume that the network is aware and a miner will reintegrate it. So it's not such a big issue, but in than scenario maybe we should add a comment here or in `AddTransactions` to highlight tht this scenario was thought of and why it works ok.

obscuren replied:
I had been thinking about this and move the locking on a per `case` basis

karalabe commented on core/transaction_pool.go:
```diff
@@ -45,10 +46,21 @@ var (
+		addedAt   time.Time
+		sentAt    time.Time
+		postTries int // amount of tx posts
+	}
```
I'm almost certain the code will contain race conditions: all the fields are accessed directly, and although most are write once, the `sentAt` (and probably `postTries` also eventually) is both read https://github.com/ethereum/go-ethereum/pull/1669/files#diff-d37f7abd23589257d36b8fa71ac686faR145 and written https://github.com/ethereum/go-ethereum/pull/1669/files#diff-d37f7abd23589257d36b8fa71ac686faR307 concurrently. I think it would be advisable to make this struct thread safe.

fjl replied:
Moving the field writes out of the post goroutine would also be sufficient.

karalabe commented:
All in all, I'm not entirely clear how the resending part of the strategy works and why is it correct or needed.

1. If I initiate a transaction, and it never gets integrated into the blockchain, then I will still have it available in my transaction pool, so whenever a new peer arrives, I'll try to propagate. Old peers currently won't accept it as they've already seen it, and if they saw it and it's still not integrated, what's the probability that sending them again will help?

2. If I initiate a transaction and it does successfully get integrated into the blockchain (i.e. dropped from my and everyone else's pool), and only afterwards disappears due to a chain reorg, then imho putting it back into the transaction pool should be enough, without re-propagating it. Any new connection will receive it upon handshake, any old connection will itself reinsert into their own pool, so there's no benefit in sending them again, and miners will try to mine it again as they include in the pool once again.

The only scenario IMHO where resending can help, is if I roll back a transaction due to reorg and then want to make sure that if some of my peers didn't get it, then they will also know about it (i.e. sanity procedure). This one however is not covered by the code because only owned transactions get resent, which is set to `false` on rollbacks.

obscuren commented:
@karalabe The resending property of this PR could actually be dropped in favour of the miner picking it up.

obscuren commented:
@caktux: This PR fails due to some Xcode failure. Any clue?

obscuren commented:
This PR has been updated. Removed the resending commit and left in the chain reorg transaction reinsertion and updated the tests per @karalabe suggestion.

karalabe commented on core/chain_manager_test.go:
```diff
@@ -483,22 +485,118 @@ func TestInsertNonceError(t *testing.T) {
+		if GetReceipt(db, tx.Hash()) != nil {
+			t.Errorf("drop %d: receipt found while shouldn't have been", i)
+		}
+	}
```
I'm thinking that maybe along these tests we should also check if the removed transactions indeed ended up back in the transaction pool. That side effect should definitely be tested somehow/somewhere.

obscuren replied:
Agreed but not in this PR. Thus PR doesn't quite care about it (tho we should make sure the event is fired).

obscuren commented on core/transaction_pool_test.go:
```diff
@@ -238,3 +238,15 @@ func TestNonceRecovery(t *testing.T) {
+	from, _ := tx.From()
+	pool.currentState().AddBalance(from, big.NewInt(1000000000000))
+	pool.eventMux.Post(RemovedTransactionEvent{types.Transactions{tx}})
+	pool.eventMux.Post(ChainHeadEvent{nil})
```
@fjl I had to post this event in order for the `transaction_pool` to receive the `RemovedTransactionEvent`. The `RemovedTransactionEvent` wasn't fired until this one was posted.`Post` is supposed to be blocking, iirc. Any clue?

obscuren commented:
Added `RemovedTransactionEvent` test and changed locking around the `TransactionPool` events.

Gustav-Simonsson commented on core/chain_manager_test.go:
```diff
@@ -483,19 +485,130 @@ func TestInsertNonceError(t *testing.T) {
+
+	// Create three transactions that will be added in the forked chain:
+	//  - pastAdd:   transaction added before the reorganiztion is detected
+	//  - freshAdd:  transaction added at the exact block the reorg is detected
```
English meta-grammar nitpicking: how about past, present, future? :) The future can be quite fresh, after all.

obscuren replied:
You need a :punch:

karalabe replied:
Second that :P

Gustav-Simonsson commented on core/chain_manager_test.go:
```diff
@@ -483,19 +485,130 @@ func TestInsertNonceError(t *testing.T) {
+	evmux := &event.TypeMux{}
+	chainman, _ := NewChainManager(db, FakePow{}, evmux)
+	chainman.SetProcessor(NewBlockProcessor(db, FakePow{}, chainman, evmux))
+	if _, err := chainman.InsertChain(chain); err != nil {
```
Let's match out block number and add to error - will make debugging of reorgs easier if these test ever fail.

karalabe commented on core/chain_manager_test.go:
```diff
@@ -483,19 +485,130 @@ func TestInsertNonceError(t *testing.T) {
+func (pow failpow) GetHashrate() int64 {
+	return 0
+}
+func (pow failpow) Turbo(bool) {
```
Could you clean these up (i.e. delete?) It was something removed in my POW PR (you have a failPow and a delayPow included in that), I guess GitHub figured that it should re-add it now that it's rebasing on top of it.

obscuren replied:
Thanks. I was going to ask you what was wrong with this. I got some really strange rebase conflict. I'll move this

karalabe commented:
Beside my above comment to delete that stale block of code, LGTM :+1: Of course maybe we should try and test this out properly on either the live net or maybe it would be a good time to re-launch the test net.
